File: php/src/ast/expressions/IndexExpression.php
<?php

readonly class IndexExpression implements Expression {

    public function __construct(
        public Token $token,
        public Expression $left,
        public Expression $index,
    ) {
    }

    public function expressionNode(): void {
    }

    public function tokenLiteral(): string {
        return "[";
    }

    public function __toString(): string {
        return "({$this->left}[{$this->index}])";
    }
}

File: java/monkey/src/main/java/root/ast/expressions/IndexExpression.java
package root.ast.expressions;

import root.LocalizedToken;

public class IndexExpression extends Expression {

    private final Expression left;
    private final Expression index;

    public IndexExpression(LocalizedToken token, Expression left, Expression index) {
        super(token);
        this.left = left;
        this.index = index;
    }

    public Expression getLeft() {
        return left;
    }

    public Expression getIndex() {
        return index;
    }

    @Override
    public String stringRep() {
        return "(%s[%s])".formatted(left.stringRep(), index.stringRep());
    }
}
